File: LeetCode/Java/src/solutions/p2348_NumberofZeroFilledSubarrays/P2348.java
package solutions.p2348_NumberofZeroFilledSubarrays;

public class P2348 {
    public long zeroFilledSubarray(int[] nums) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 3 ms @ (beats) 99.69%
         * Memory Usage: 60.1 MB @ (beats) 27.27%
         */
        // long ans = 0;
        // long sub = 0;
        //
        // for (int n : nums) {
        //     if (n == 0) {
        //         sub++;
        //     } else {
        //         sub = 0;
        //     }
        //     ans += sub;
        // }
        // return ans;

        /*
         * Solution v2: Two pointer
         *
         * Runtime: 3 ms @ (beats) 99.69%
         * Memory Usage: 59.9 MB @ (beats) 41.69%
         */
        long ans = 0;
        int l = nums.length;
        int j = -1;
        for (int i = 0; i < l; i++) {
            if (nums[i] == 0) {
                ans += i - j;
            } else {
                j = i;
            }
        }
        return ans;
    }
}

File: LeetCode/Python/solutions/p2348.py
from typing import List


class Solution:
    def zeroFilledSubarray(self, nums: List[int]) -> int:

        #
        # Solution v1: Brute Force
        #
        # Runtime: 1090 ms @ (beats) 76.32%
        # Memory Usage: 24.5 MB @ (beats) 78.9%
        #
        # ans = 0
        #
        # sub = 0
        # for i in range(len(nums)):
        #     if nums[i]:
        #         ans += sum(range(1, sub + 1))
        #         sub = 0
        #     else:
        #         sub += 1
        #
        # ans += sum(range(1, sub + 1))
        #
        # return ans

        #
        # Solution v1.5: Brute Force
        # Simplify Algorithm
        #
        # Runtime: 1069 ms @ (beats) 85.87%
        # Memory Usage: 24.7 MB @ (beats) 30.39%
        #
        ans = 0
        sub = 0
        for n in nums:
            # increment the counter if nums[i]= 0
            if n:
                sub = 0
            else:
                sub += 1
            # accumulate counter to answer for each element
            ans += sub

        return ans


if __name__ == "__main__":
    solution = Solution()

    nums = [1, 3, 0, 0, 2, 0, 0, 4]
    print(f"return: {solution.zeroFilledSubarray(nums)}")  # 6

    nums = [0, 0, 0, 2, 0, 0]
    print(f"return: {solution.zeroFilledSubarray(nums)}")  # 9

    nums = [2, 10, 2019]
    print(f"return: {solution.zeroFilledSubarray(nums)}")  # 0
